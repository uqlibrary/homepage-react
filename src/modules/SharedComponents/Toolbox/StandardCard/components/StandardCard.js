import React, { Component } from 'react';
import PropTypes from 'prop-types';

import Card from '@material-ui/core/Card';
import CardContent from '@material-ui/core/CardContent';
import CardHeader from '@material-ui/core/CardHeader';
import { HelpIcon } from '../../HelpDrawer';
import { withStyles } from '@material-ui/core/styles';

export const styles = theme => ({
    card: {
        overflow: 'unset',
        fontWeight: theme.typography.fontWeightRegular,
    },
    cardContentNoPadding: {
        paddingTop: '0px !important',
        paddingBottom: '0px !important',
        paddingLeft: '0px !important',
        paddingRight: '0px !important',
    },
    cardHeaderPrimary: {
        color: theme.palette.white.main,
        overflow: 'hidden',
        whiteSpace: 'nowrap',
        textOverflow: 'ellipsis',
        backgroundColor: theme.palette.primary.main,
        borderRadius: '4px 4px 0px 0px',
        padding: '12px 24px',
    },
    cardHeaderAccent: {
        color: theme.palette.white.main,
        overflow: 'hidden',
        whiteSpace: 'nowrap',
        textOverflow: 'ellipsis',
        backgroundColor: theme.palette.accent.main,
        borderRadius: '4px 4px 0px 0px',
        padding: '12px 24px',
    },
    fullHeight: {
        height: '100%',
    },
});

export class Cards extends Component {
    static propTypes = {
        title: PropTypes.any,
        primaryHeader: PropTypes.bool,
        accentHeader: PropTypes.bool,
        fullHeight: PropTypes.bool,
        noPadding: PropTypes.bool,
        noHeader: PropTypes.bool,
        children: PropTypes.any,
        classes: PropTypes.object.isRequired,
        help: PropTypes.object,
        customBackgroundColor: PropTypes.any,
        customTitleColor: PropTypes.any,
        customTitleBgColor: PropTypes.any,
        customTextColor: PropTypes.any,
        squareTop: PropTypes.bool,
        smallTitle: PropTypes.bool,
        standardCardId: PropTypes.string,
        subCard: PropTypes.bool,
        style: PropTypes.object,
    };

    render() {
        const {
            classes,
            title,
            help,
            children,
            primaryHeader,
            accentHeader,
            smallTitle = false,
            subCard = false,
            style = {},
        } = this.props;
        const customBG = !!this.props.customBackgroundColor
            ? { backgroundColor: this.props.customBackgroundColor }
            : null;
        const customTitleBG = !!this.props.customTitleBgColor
            ? { backgroundColor: this.props.customTitleBgColor }
            : null;
        const customTitle = !!this.props.customTitleColor ? { color: this.props.customTitleColor } : null;
        const customText = !!this.props.customTextColor ? { color: `${this.props.customTextColor} !important` } : null;
        const fullHeight = !!this.props.fullHeight ? { height: '100%' } : null;
        const squareTop = !!this.props.squareTop
            ? { borderTopLeftRadius: 0, borderTopRightRadius: 0 }
            : { borderTopLeftRadius: 4, borderTopRightRadius: 4 };
        const standardCardId = !!this.props.standardCardId
            ? this.props.standardCardId
            : `standard-card${typeof title === 'string' ? '-' + title.replace(/ /g, '-').toLowerCase() : ''}`;
        return (
            <Card
                data-testid={standardCardId}
                id={standardCardId}
                className={`${classes.card} StandardCard`}
                style={{ ...customBG, ...fullHeight, ...style }}
            >
                {!this.props.noHeader && (
                    <CardHeader
                        style={{ ...squareTop, ...customTitleBG, ...customTitle }}
                        title={title}
                        titleTypographyProps={{
                            variant: smallTitle ? 'h6' : 'h5',
                            component: subCard ? 'h4' : 'h3',
                            color: 'inherit',
                            'data-testid': `${standardCardId}-header`,
                        }}
                        action={!!help && !!help.text && <HelpIcon {...help} />}
                        classes={{
                            root:
                                (primaryHeader && classes.cardHeaderPrimary) ||
                                (accentHeader && classes.cardHeaderAccent),
                        }}
                    />
                )}
                <CardContent
                    data-testid={`${standardCardId}-content`}
                    className={`${(this.props.noPadding && classes.cardContentNoPadding) || ''}`}
                    style={{ ...customText }}
                >
                    {children}
                </CardContent>
            </Card>
        );
    }
}

const StyledCard = withStyles(styles, { withTheme: true })(Cards);
export const StandardCard = props => <StyledCard {...props} />;
export default StandardCard;
