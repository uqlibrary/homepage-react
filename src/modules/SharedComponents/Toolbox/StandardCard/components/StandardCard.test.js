import { StandardCard, Cards } from './StandardCard';

function setup(testProps, isShallow = true) {
    const props = {
        title: 'card title',
        classes: {
            card: 'testClass',
        },
        ...testProps,
    };
    return getElement(Cards, props, isShallow);
}

describe('Cards component', () => {
    it('renders with title and no help icon', () => {
        const wrapper = setup({});
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('renders with title and help button', () => {
        const wrapper = setup({
            help: {
                title: 'help',
                text: 'help text',
                buttonLabel: 'OK',
            },
        });
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('renders with custom colours and full height', () => {
        const wrapper = setup({
            customBackgroundColor: '#fcc',
            customTitleColor: '#111',
            customTitleBgColor: '#ccc',
            fullHeight: true,
            noPadding: true,
            primaryHeader: true,
            classes: {
                cardHeaderPurple: 'purple',
                cardContentNoPadding: 'no-padding',
                cardHeaderPrimary: '#555',
            },
        });
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('renders with square top and accent header', () => {
        const wrapper = setup({
            squareTop: true,
            accentHeader: true,
            classes: {
                cardHeaderAccent: '#333',
            },
        });

        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('renders with small title and as a subcard', () => {
        const wrapper = setup({
            smallTitle: true,
            subCard: true,
        });
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('renders given ID for standar card', () => {
        const wrapper = setup({ standardCardId: 'test-card' });
        expect(toJson(wrapper)).toMatchSnapshot();
    });
});

describe('StandardCard component', () => {
    it('should render StyledCard with same props', () => {
        const wrapper = getElement(
            StandardCard,
            {
                test1: 'test1value',
                test2: 'test2value',
            },
            true,
        );
        expect(toJson(wrapper)).toMatchSnapshot();
    });
});
