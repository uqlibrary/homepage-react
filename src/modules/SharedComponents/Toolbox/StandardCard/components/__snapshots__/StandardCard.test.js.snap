// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cards component renders given ID for standar card 1`] = `
<WithStyles(ForwardRef(Card))
  className="testClass StandardCard "
  data-analyticsid="test-card"
  data-testid="test-card"
  id="test-card"
  style={Object {}}
  variant="elevation"
>
  <WithStyles(ForwardRef(CardHeader))
    action={<React.Fragment />}
    classes={
      Object {
        "root": undefined,
      }
    }
    style={
      Object {
        "borderTopLeftRadius": 4,
        "borderTopRightRadius": 4,
      }
    }
    title="card title"
    titleTypographyProps={
      Object {
        "color": "inherit",
        "component": "h2",
        "data-testid": "test-card-header",
        "variant": "h5",
      }
    }
  />
  <WithStyles(ForwardRef(CardContent))
    className=""
    data-testid="test-card-content"
    style={Object {}}
  />
</WithStyles(ForwardRef(Card))>
`;

exports[`Cards component renders with custom colours and full height 1`] = `
<WithStyles(ForwardRef(Card))
  className="undefined StandardCard "
  data-analyticsid="standard-card-card-title"
  data-testid="standard-card-card-title"
  id="standard-card-card-title"
  style={
    Object {
      "backgroundColor": "#fcc",
      "height": "100%",
    }
  }
  variant="elevation"
>
  <WithStyles(ForwardRef(CardHeader))
    action={<React.Fragment />}
    classes={
      Object {
        "root": "#555",
      }
    }
    style={
      Object {
        "backgroundColor": "#ccc",
        "borderTopLeftRadius": 4,
        "borderTopRightRadius": 4,
        "color": "#111",
      }
    }
    title="card title"
    titleTypographyProps={
      Object {
        "color": "inherit",
        "component": "h2",
        "data-testid": "standard-card-card-title-header",
        "variant": "h5",
      }
    }
  />
  <WithStyles(ForwardRef(CardContent))
    className="no-padding"
    data-testid="standard-card-card-title-content"
    style={
      Object {
        "color": "blue !important",
      }
    }
  />
</WithStyles(ForwardRef(Card))>
`;

exports[`Cards component renders with small title and as a subcard 1`] = `
<WithStyles(ForwardRef(Card))
  className="testClass StandardCard "
  data-analyticsid="standard-card-card-title"
  data-testid="standard-card-card-title"
  id="standard-card-card-title"
  style={Object {}}
  variant="elevation"
>
  <WithStyles(ForwardRef(CardHeader))
    action={<React.Fragment />}
    classes={
      Object {
        "root": undefined,
      }
    }
    style={
      Object {
        "borderTopLeftRadius": 4,
        "borderTopRightRadius": 4,
      }
    }
    title="card title"
    titleTypographyProps={
      Object {
        "color": "inherit",
        "component": "h3",
        "data-testid": "standard-card-card-title-header",
        "variant": "h6",
      }
    }
  />
  <WithStyles(ForwardRef(CardContent))
    className=""
    data-testid="standard-card-card-title-content"
    style={Object {}}
  />
</WithStyles(ForwardRef(Card))>
`;

exports[`Cards component renders with square top and accent header 1`] = `
<WithStyles(ForwardRef(Card))
  className="undefined StandardCard "
  data-analyticsid="standard-card-card-title"
  data-testid="standard-card-card-title"
  id="standard-card-card-title"
  style={Object {}}
  variant="elevation"
>
  <WithStyles(ForwardRef(CardHeader))
    action={<React.Fragment />}
    classes={
      Object {
        "root": "#333",
      }
    }
    style={
      Object {
        "borderTopLeftRadius": 0,
        "borderTopRightRadius": 0,
      }
    }
    title="card title"
    titleTypographyProps={
      Object {
        "color": "inherit",
        "component": "h2",
        "data-testid": "standard-card-card-title-header",
        "variant": "h5",
      }
    }
  />
  <WithStyles(ForwardRef(CardContent))
    className=""
    data-testid="standard-card-card-title-content"
    style={Object {}}
  />
</WithStyles(ForwardRef(Card))>
`;

exports[`Cards component renders with title and help button 1`] = `
<WithStyles(ForwardRef(Card))
  className="testClass StandardCard "
  data-analyticsid="standard-card-card-title"
  data-testid="standard-card-card-title"
  id="standard-card-card-title"
  style={Object {}}
  variant="elevation"
>
  <WithStyles(ForwardRef(CardHeader))
    action={<React.Fragment />}
    classes={
      Object {
        "root": undefined,
      }
    }
    style={
      Object {
        "borderTopLeftRadius": 4,
        "borderTopRightRadius": 4,
      }
    }
    title="card title"
    titleTypographyProps={
      Object {
        "color": "inherit",
        "component": "h2",
        "data-testid": "standard-card-card-title-header",
        "variant": "h5",
      }
    }
  />
  <WithStyles(ForwardRef(CardContent))
    className=""
    data-testid="standard-card-card-title-content"
    style={Object {}}
  />
</WithStyles(ForwardRef(Card))>
`;

exports[`Cards component renders with title and no help icon 1`] = `
<WithStyles(ForwardRef(Card))
  className="testClass StandardCard "
  data-analyticsid="standard-card-card-title"
  data-testid="standard-card-card-title"
  id="standard-card-card-title"
  style={Object {}}
  variant="elevation"
>
  <WithStyles(ForwardRef(CardHeader))
    action={<React.Fragment />}
    classes={
      Object {
        "root": undefined,
      }
    }
    style={
      Object {
        "borderTopLeftRadius": 4,
        "borderTopRightRadius": 4,
      }
    }
    title="card title"
    titleTypographyProps={
      Object {
        "color": "inherit",
        "component": "h2",
        "data-testid": "standard-card-card-title-header",
        "variant": "h5",
      }
    }
  />
  <WithStyles(ForwardRef(CardContent))
    className=""
    data-testid="standard-card-card-title-content"
    style={Object {}}
  />
</WithStyles(ForwardRef(Card))>
`;

exports[`StandardCard component should render StyledCard with same props 1`] = `
<WithStyles(Cards)
  test1="test1value"
  test2="test2value"
/>
`;
